dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/validation/TranslationsCheck.java

```diff
@@ -0,0 +1,131 @@
+{
+    @Override
+    public <T extends IdentifiableObject> void check( ObjectBundle bundle, Class<T> klass, List<T> persistedObjects,
+        List<T> nonPersistedObjects, ImportStrategy importStrategy, ValidationContext context,
+        Consumer<ObjectReport> addReports )
+    {
+        List<T> objects = selectObjects( persistedObjects, nonPersistedObjects, importStrategy );
+
+        if ( CollectionUtils.isEmpty( objects ) )
+        {
+            return;
+        }
+
+        Schema schema = context.getSchemaService().getDynamicSchema( klass );
+
+        for ( int i = 0; i < objects.size(); i++ )
+        {
+            run( objects.get( i ), klass, addReports, schema, i );
+        }
+    }
+
+    private <T extends IdentifiableObject> void run( IdentifiableObject object, Class<T> klass,
+        Consumer<ObjectReport> addReports, Schema schema, int index )
+    {
+        Set<Translation> translations = object.getTranslations();
+
+        if ( CollectionUtils.isEmpty( translations ) )
+        {
+            return;
+        }
+
+        ObjectReport objectReport = new ObjectReport( klass, index );
+
+        if ( !schema.isTranslatable() )
+        {
+            objectReport.addErrorReport(
+                new ErrorReport( Translation.class, ErrorCode.E1107, klass.getSimpleName() ).setErrorKlass( klass ) );
+            addReports.accept( objectReport );
+            return;
+        }
+
+        Map<String, String> mapPropertyLocale = new HashMap<>();
+
+        for ( Translation translation : translations )
+        {
+            if ( mapPropertyLocale.containsKey( translation.getProperty() )
+                && mapPropertyLocale.get( translation.getProperty() ).equals( translation.getLocale() ) )
+            {
+                objectReport.addErrorReport(
+                    new ErrorReport( Translation.class, ErrorCode.E1106, translation.getProperty(),
+                        translation.getLocale() )
+                            .setErrorKlass( klass ) );
+            }
+            else
+            {
+                mapPropertyLocale.put( translation.getProperty(), translation.getLocale() );
+            }
+
+            if ( translation.getLocale() == null )
+            {
+                objectReport.addErrorReport(
+                    new ErrorReport( Translation.class, ErrorCode.E4000, "locale" ).setErrorKlass( klass ) );
+            }
+
+            if ( translation.getProperty() == null )
+            {
+                objectReport.addErrorReport( new ErrorReport( Translation.class, ErrorCode.E4000, "property" )
+                    .setErrorKlass( klass ) );
+            }
+
+            if ( translation.getValue() == null )
+            {
+                objectReport.addErrorReport(
+                    new ErrorReport( Translation.class, ErrorCode.E4000, "value" ).setErrorKlass( klass ) );
+            }
+
+            addReports.accept( objectReport );
```

Comment:
Do we always want to add the report even if there is no error?

Reply:
no, i added an empty check.